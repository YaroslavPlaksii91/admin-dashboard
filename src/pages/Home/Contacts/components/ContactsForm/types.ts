import { ContactType } from '../ContactsItem/types';

export type AddContactsData = {
  firstName: string;
  lastName: string;
  email: string;
  address: string;
  date: string;
  photo: Blob[];
};

export type UpdateContactData = {
  name: string;
  email: string;
  address: string;
  image: string;
  date: string;
};

export type ContactsFormProps = {
  addContact: (data: AddContactsData) => void;
  editContact: (data: AddContactsData) => void;
  activeContact?: ContactType;
};
