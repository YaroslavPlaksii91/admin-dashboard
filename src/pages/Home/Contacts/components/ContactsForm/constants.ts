import { IMAGE_SIZE } from '@components/AddPhoto/constants';

export const CONTACTS_FORM_FIELDS = Object.freeze({
  PHOTO: 'photo',
  FIRST_NAME: 'firstName',
  LAST_NAME: 'lastName',
  EMAIL: 'email',
  ADDRESS: 'address',
});

export const CONTACTS_FORM_CONFIG = Object.freeze({
  [CONTACTS_FORM_FIELDS.PHOTO]: {
    label: 'Add photo',
    name: 'photo',
    size: IMAGE_SIZE.BIG,
  },
  [CONTACTS_FORM_FIELDS.FIRST_NAME]: {
    label: 'first name',
    type: 'text',
    name: 'firstName',
    placeholder: 'First name',
  },
  [CONTACTS_FORM_FIELDS.LAST_NAME]: {
    label: 'last name',
    type: 'text',
    name: 'lastName',
    placeholder: 'Last name',
  },
  [CONTACTS_FORM_FIELDS.EMAIL]: {
    label: 'email',
    type: 'email',
    name: 'email',
    placeholder: 'Email',
  },
  [CONTACTS_FORM_FIELDS.ADDRESS]: {
    label: 'address',
    type: 'text',
    name: 'address',
    placeholder: 'Address',
  },
});
