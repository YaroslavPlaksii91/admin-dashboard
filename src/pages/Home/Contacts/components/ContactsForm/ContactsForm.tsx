import { FC } from 'react';
import { useForm, SubmitHandler } from 'react-hook-form';
import { Button, Box } from '@mui/material';

import { FormInput } from '@components/FormInput/FormInput';
import { AddPhoto } from '@components/AddPhoto/AddPhoto';
import { EMAIL_REGEX } from '@utils/constants';

import { CONTACTS_FORM_CONFIG, CONTACTS_FORM_FIELDS } from './constants';
import { AddContactsData, ContactsFormProps } from './types';

export const ContactsForm: FC<ContactsFormProps> = ({
  addContact,
  editContact,
  activeContact,
}) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<AddContactsData>();

  const firstNameDefault = activeContact?.name.split(' ')[0];
  const lastNameDefault = activeContact?.name.split(' ')[1];

  const onSubmit: SubmitHandler<AddContactsData> = data => {
    if (activeContact) {
      editContact({ ...data, date: activeContact.date });
    } else {
      addContact(data);
    }
  };

  return (
    <Box
      component="form"
      onSubmit={handleSubmit(onSubmit)}
      noValidate
      sx={{ margin: '32px 0 24px' }}
    >
      <AddPhoto
        {...CONTACTS_FORM_CONFIG[CONTACTS_FORM_FIELDS.PHOTO]}
        register={register('photo', {
          required: 'Choose a photo',
        })}
        errors={errors}
        src={activeContact?.image}
      />

      <FormInput
        {...CONTACTS_FORM_CONFIG[CONTACTS_FORM_FIELDS.FIRST_NAME]}
        register={register('firstName', {
          required: 'First name is required',
          minLength: {
            value: 2,
            message: 'First name must be at least 2 characters',
          },
        })}
        errors={errors}
        defaultValue={firstNameDefault}
      />

      <FormInput
        {...CONTACTS_FORM_CONFIG[CONTACTS_FORM_FIELDS.LAST_NAME]}
        register={register('lastName', {
          required: 'Last name is required',
          minLength: {
            value: 2,
            message: 'Last name must be at least 2 characters',
          },
        })}
        errors={errors}
        defaultValue={lastNameDefault}
      />

      <FormInput
        {...CONTACTS_FORM_CONFIG[CONTACTS_FORM_FIELDS.EMAIL]}
        register={register('email', {
          required: 'Email is required',
          pattern: {
            value: EMAIL_REGEX,
            message: 'Invalid email address',
          },
        })}
        errors={errors}
        defaultValue={activeContact?.email}
      />

      <FormInput
        {...CONTACTS_FORM_CONFIG[CONTACTS_FORM_FIELDS.ADDRESS]}
        register={register('address', {
          required: 'Address is required',
          minLength: {
            value: 10,
            message: 'Address must be at least 10 characters',
          },
        })}
        errors={errors}
        defaultValue={activeContact?.address}
      />

      <Button variant="contained" type="submit" fullWidth>
        Save
      </Button>
    </Box>
  );
};
