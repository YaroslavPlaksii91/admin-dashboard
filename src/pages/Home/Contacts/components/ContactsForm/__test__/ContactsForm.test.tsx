import { cleanup, render, screen, fireEvent } from '@testing-library/react';

import { ContactsForm } from '../ContactsForm';

describe('ContactsForm component', () => {
  afterEach(cleanup);

  const mockAddContact = jest.fn();
  const mockEditContact = jest.fn();

  const defaultProps = {
    addContact: mockAddContact,
    editContact: mockEditContact,
  };

  it('should be rendered', () => {
    const { asFragment } = render(<ContactsForm {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    render(<ContactsForm {...defaultProps} />);

    expect(screen.getByRole('button', { name: 'Save' })).toBeInTheDocument();
    expect(screen.getByLabelText('first name')).toBeInTheDocument();
    expect(screen.getByLabelText('last name')).toBeInTheDocument();
    expect(screen.getByLabelText('email')).toBeInTheDocument();
    expect(screen.getByLabelText('address')).toBeInTheDocument();
  });

  it('handles form submission correctly for adding', () => {
    render(<ContactsForm {...defaultProps} />);

    fireEvent.input(screen.getByLabelText('first name'), {
      target: { value: 'John' },
    });
    fireEvent.input(screen.getByLabelText('last name'), {
      target: { value: 'Doe' },
    });
    fireEvent.input(screen.getByLabelText('email'), {
      target: { value: 'johndoe@example.com' },
    });
    fireEvent.input(screen.getByLabelText('address'), {
      target: { value: '123 Main St' },
    });

    fireEvent.click(screen.getByRole('button', { name: 'Save' }));
  });

  it('requires a photo to be selected', () => {
    render(<ContactsForm {...defaultProps} />);

    fireEvent.input(screen.getByLabelText('first name'), {
      target: { value: 'John' },
    });
    fireEvent.input(screen.getByLabelText('last name'), {
      target: { value: 'Doe' },
    });
    fireEvent.input(screen.getByLabelText('email'), {
      target: { value: 'johndoe@example.com' },
    });
    fireEvent.input(screen.getByLabelText('address'), {
      target: { value: '123 Main St' },
    });

    fireEvent.click(screen.getByRole('button', { name: 'Save' }));

    expect(defaultProps.addContact).not.toHaveBeenCalled();
    expect(defaultProps.editContact).not.toHaveBeenCalled();
  });
});
