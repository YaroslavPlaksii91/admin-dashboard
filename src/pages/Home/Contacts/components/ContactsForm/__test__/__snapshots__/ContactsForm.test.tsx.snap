// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsForm component should be rendered 1`] = `
<DocumentFragment>
  <form
    class="MuiBox-root css-1a2bv8d"
    novalidate=""
  >
    <div
      class="MuiBox-root css-14h66l3"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-g4d8h0-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="photo"
            name="photo"
            type="file"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <label
        class="MuiFormLabel-root MuiFormLabel-colorPrimary css-goj8zy-MuiFormLabel-root"
        for="photo"
      >
        <span
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1nntmli-MuiButtonBase-root-MuiButton-root"
          role="button"
          tabindex="0"
        >
          <svg
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1xzf3w3-MuiSvgIcon-root"
            focusable="false"
            role="img"
            viewBox="0 0 24 24"
          >
            <use
              href="/src/assets/icons/sprite.svg#icon-addPhoto"
            />
            <title>
              Add photo
            </title>
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        Add photo
      </label>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r1:"
        id=":r1:-label"
      >
        first name
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r1:"
          name="firstName"
          placeholder="First name"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              first name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r2:"
        id=":r2:-label"
      >
        last name
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r2:"
          name="lastName"
          placeholder="Last name"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              last name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r3:"
        id=":r3:-label"
      >
        email
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r3:"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              email
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r4:"
        id=":r4:-label"
      >
        address
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r4:"
          name="address"
          placeholder="Address"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              address
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      Save
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
